AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/RemoteSyncEngine.swift

```diff
@@ -228,18 +236,100 @@ class RemoteSyncEngine: RemoteSyncEngineBehavior {
+            if let api = self.api, let storageAdapter = self.storageAdapter {
+                stateMachine.notify(action: .authenticationNotRequired(api, storageAdapter))
+            }
+            return
+        }
+
+        guard let auth = auth else {
+            let dataStoreError = DataStoreError.configuration("Missing auth with auth enabled models",
+                                                              "Configure Amplify.Auth")
+            stateMachine.notify(action: .errored(dataStoreError))
+            return
+        }
+
+        isSignedIn(auth: auth) { result in
+            switch result {
+            case .success(let isSignedIn):
+                if isSignedIn, let api = self.api, let storageAdapter = self.storageAdapter {
+                    self.stateMachine.notify(action: .authenticatedUser(api, storageAdapter, auth))
+                } else {
+                    self.stateMachine.notify(action: .requireAuthenticatedUser)
+                }
+            case .failure(let authError):
+                self.stateMachine.notify(action: .errored(authError))
+            }
+        }
+    }
+
+    private func waitForAuthenticatedUser() {
+        log.debug(#function)
+        signInListener = Amplify.Hub.listen(to: .auth) { payload in
+            guard payload.eventName == HubPayload.EventName.Auth.signIn else {
+                return
+            }
+
+            guard let auth = self.auth else {
+                let dataStoreError = DataStoreError.configuration("Missing auth with auth enabled models",
+                                                                  "Configure Amplify.Auth")
+                self.stateMachine.notify(action: .errored(dataStoreError))
+                return
+            }
+
+            self.isSignedIn(auth: auth) { result in
+                switch result {
+                case .success(let isSignedIn):
+                    if isSignedIn, let api = self.api, let storageAdapter = self.storageAdapter {
+                        self.removeSignInListener()
+                        self.stateMachine.notify(action: .authenticatedUser(api, storageAdapter, auth))
+                    }
+                case .failure(let authError):
+                    self.stateMachine.notify(action: .errored(authError))
+                }
+            }
+        }
+    }
+
+    private func removeSignInListener() {
+        if let listener = signInListener {
+            Amplify.Hub.removeListener(listener)
+        }
+    }
+
+    private func isSignedIn(auth: AuthCategoryBehavior, onComplete: @escaping (Result<Bool, AuthError>) -> Void) {
+        _ = auth.fetchAuthSession(options: nil) { event in
+            switch event {
+            case .success(let authSession):
+                if authSession.isSignedIn {
+                    onComplete(.success(true))
+                } else {
+                    onComplete(.success(false))
+                }
```

Comment:
could be simplified to `onComplete(.success(authSession.isSignedIn))`

Reply:
nice, thanks!